libr/bin/format/elf/elf.c

```diff
@@ -1236,314 +1238,349 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
+	if (!rel_sec || rel_sec->size < 1) {
+		return NULL;
 	}
-	nrel = (ut32)((int)rel_sec->size / (int)tsize);
+
+	nrel = (ut32) ((int)rel_sec->size / (int)tsize);
 	if (nrel < 1) {
-		return -1;
-	}
-	if (is_rela) {
-		rela = calloc (nrel, tsize);
-		if (!rela) {
-			return -1;
-		}
-	} else {
-		rel = calloc (nrel, tsize);
-		if (!rel) {
-			return -1;
-		}
+		return NULL;
 	}
+
+	rel_cache = ht_up_new_size (nrel, NULL, rel_cache_free, NULL);
+
 	for (j = k = 0; j < rel_sec->size && k < nrel; j += tsize, k++) {
-		int l = 0;
 		if (rel_sec->offset + j > bin->size) {
 			goto out;
 		}
 		if (rel_sec->offset + j + tsize > bin->size) {
 			goto out;
 		}
-		len = r_buf_read_at (
-			bin->b, rel_sec->offset + j, is_rela ? rla : rl,
-			is_rela ? sizeof (Elf_ (Rela)) : sizeof (Elf_ (Rel)));
+		int len = r_buf_read_at (bin->b, rel_sec->offset + j, rl, tsize);
 		if (len < 1) {
 			goto out;
 		}
-#if R_BIN_ELF64
-		if (is_rela) {
-			rela[k].r_offset = READ64 (rla, l);
-			rela[k].r_info = READ64 (rla, l);
-			rela[k].r_addend = READ64 (rla, l);
-		} else {
-			rel[k].r_offset = READ64 (rl, l);
-			rel[k].r_info = READ64 (rl, l);
+
+		struct ht_rel_t *rel = R_NEW0 (struct ht_rel_t);
+		if (!rel) {
+			goto out;
 		}
-#else
-		if (is_rela) {
-			rela[k].r_offset = READ32 (rla, l);
-			rela[k].r_info = READ32 (rla, l);
-			rela[k].r_addend = READ32 (rla, l);
+
+		rel->is_rela = bin->is_rela == DT_RELA;
+		rel->k = k;
+		read_ht_rel (bin, rel, rl);
+		ht_up_insert (rel_cache, REL_SYM, rel);
+	}
+	return rel_cache;
+out:
+	ht_up_free (rel_cache);
+	return NULL;
+}
+
+static ut64 get_import_addr(ELFOBJ *bin, int sym) {
+	ut64 plt_sym_addr = -1;
+	ut64 got_addr, got_offset, plt_addr;
+	int k, len, nrel;
+
+	if ((!bin->shdr || !bin->strtab) && !bin->phdr) {
+		return -1;
+	}
+
+	// create rel/rela cache if not already there
+	if (!bin->rel_cache) {
+		bin->rel_cache = rel_cache_new (bin);
+	}
+
+	// lookup the right rel/rela entry
+	struct ht_rel_t *rel = ht_up_find (bin->rel_cache, sym, NULL);
+	if (!rel) {
+		return -1;
+	}
+
+	nrel = bin->rel_cache->count;
+	if ((got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got")) == -1 &&
+	    (got_offset = Elf_(r_bin_elf_get_section_offset) (bin, ".got.plt")) == -1) {
+		return -1;
+	}
+	if ((got_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".got")) == -1 &&
+	    (got_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".got.plt")) == -1) {
+		return -1;
+	}
+	RBinElfSection *plt_section = get_section_by_name (bin, ".plt");
+
+	int of = REL_OFFSET - got_addr + got_offset;
+	switch (bin->ehdr.e_machine) {
+	case EM_PPC:
+	case EM_PPC64: {
+		RBinElfSection *s = plt_section;
+		if (s) {
+			ut8 buf[4] = { 0 };
+			len = r_buf_read_at (bin->b, s->offset, buf, sizeof (buf));
+			if (len < 4) {
+				goto out;
+			}
+			if (bin->endian) {
+				ut64 base = r_read_be32 (buf);
+				base -= (nrel * 16);
+				base += (rel->k * 16);
+				plt_addr = base;
+			} else {
+				// FIXME: this does not seem to work as
```

Comment:
this needs to be fixed somehow, but I know nothing about PPC64. However, as mentioned in https://github.com/radare/radare2/issues/9825#issuecomment-451498683 it seems this was introduced to fix something that is actually not fixed, so...

Reply:
ppc-linux is a bit weird in many aspects xD

Reply:
yeah :/ And I really don't know the architecture at all :(